### PR TITLE
[ML] Allow users to specify similarity field

### PR DESCRIPTION
**Pull request by jonathan-buttner**
This PR preserves the user provided `similarity` field value if specified. For OpenAI and Cohere if the user did not specify a value, then it defaults to `DOT_PRODUCT`.

For e5 if the user does not specify it defaults to `COSINE`. The e5 configuration does not allow the user to specify the dimensions. Right now they are hardcoded to 384.

This also changes how huggingface works. It allows the user to specify a value, if they don't we won't default it.

## E5 Examples

<details><summary>Defaults used</summary>

```
PUT _inference/text_embedding/test
{
  "service": "elasticsearch",
  "service_settings": {
    "model_id": ".multilingual-e5-small",
    "num_threads": 1,
    "num_allocations": 1
  }
}

Response:

{
  "model_id": "test",
  "task_type": "text_embedding",
  "service": "elasticsearch",
  "service_settings": {
    "num_allocations": 1,
    "num_threads": 1,
    "model_id": ".multilingual-e5-small",
    "similarity": "cosine",
    "dimensions": 384
  },
  "task_settings": {}
}
```

</details>

### CODE REVIEW

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/cohere/CohereService.java [hunk 1]
```diff
@@ -216,13 +216,16 @@ public void checkModelConfig(Model model, ActionListener<Model> listener) {
     }
 
     private CohereEmbeddingsModel updateModelWithEmbeddingDetails(CohereEmbeddingsModel model, int embeddingSize) {
+        var similarityFromModel = model.getServiceSettings().similarity();
+        var similarityToUse = similarityFromModel == null ? SimilarityMeasure.DOT_PRODUCT : similarityFromModel;
```
Comment (jonathan-buttner) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/cohere/CohereService.java:220]: For cohere do we want to default to dot product if it is unspecified by the user?

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/cohere/CohereServiceSettings.java [hunk 2]
```diff
@@ -110,23 +110,25 @@ public CohereServiceSettings(StreamInput in) throws IOException {
...
     }
 
-    public URI getUri() {
+    public URI uri() {
         return uri;
     }
 
-    public SimilarityMeasure getSimilarity() {
+    @Override
+    public SimilarityMeasure similarity() {
```
Comment (jonathan-buttner) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/cohere/CohereServiceSettings.java:118]: I'm switching these to be inline with the other services. We had a bug where we weren't overriding the default implementation in the `ServiceSettings` interface which defaults to returning `null`.

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/huggingface/HuggingFaceServiceSettings.java [hunk 3]
```diff
@@ -144,10 +144,12 @@ public URI uri() {
         return uri;
     }
 
+    @Override
```
Comment (jonathan-buttner) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/huggingface/HuggingFaceServiceSettings.java:147]: Just calling it out that these are overriding the default.

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/huggingface/HuggingFaceService.java [hunk 4]
```diff
@@ -61,7 +61,7 @@ public void checkModelConfig(Model model, ActionListener<Model> listener) {
     private static HuggingFaceEmbeddingsModel updateModelWithEmbeddingDetails(HuggingFaceEmbeddingsModel model, int embeddingSize) {
         var serviceSettings = new HuggingFaceServiceSettings(
             model.getServiceSettings().uri(),
-            null, // Similarity measure is unknown
+            model.getServiceSettings().similarity(), // we don't know the similarity but use whatever the user specified
```
Comment (jonathan-buttner) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/huggingface/HuggingFaceService.java:64]: I'm assuming we want to allow users for hugging face to specify the similarity as well?

File: x-pack/plugin/inference/src/test/java/org/elasticsearch/xpack/inference/services/openai/OpenAiServiceTests.java [hunk 5]
```diff
@@ -937,7 +1049,7 @@ public void testCheckModelConfig_ReturnsNewModelReference_AndDoesNotSendDimensio
                         "secret",
                         "model",
                         "user",
-                        SimilarityMeasure.DOT_PRODUCT,
+                        SimilarityMeasure.COSINE,
```
Comment (jonathan-buttner) [inline x-pack/plugin/inference/src/test/java/org/elasticsearch/xpack/inference/services/openai/OpenAiServiceTests.java:1052]: The test specifies cosine so it should stay cosine now.

**Comment by jonathan-buttner** [comment]
@elasticmachine merge upstream

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/elasticsearch/MultilingualE5SmallInternalServiceSettings.java [hunk 6]
```diff
@@ -8,42 +8,107 @@
...
     }
 
     public MultilingualE5SmallInternalServiceSettings(StreamInput in) throws IOException {
         super(in.readVInt(), in.readVInt(), in.readString());
+
+        if (in.getTransportVersion().onOrAfter(TransportVersions.ML_INFERENCE_L2_NORM_SIMILARITY_ADDED)) {
+            similarity = in.readEnum(SimilarityMeasure.class);
+            dimensions = in.readVInt();
+        } else {
+            similarity = DEFAULT_SIMILARITY;
```
Comment (jonathan-buttner) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/elasticsearch/MultilingualE5SmallInternalServiceSettings.java:62]: In the other services we've allowed these to be `null`. Seems like we want these to be something and the dimensions were hardcoded before to `384`. In a future PR we can do the same thing that the other external services do which is make an inference request to determine the dimensions dynamically.

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/elasticsearch/MultilingualE5SmallInternalServiceSettings.java [hunk 7]
```diff
@@ -8,42 +8,107 @@
...
      * @param map Source map containing the config
      * @return The {@code MultilingualE5SmallServiceSettings} builder
      */
-    public static MultilingualE5SmallInternalServiceSettings.Builder fromMap(Map<String, Object> map) {
+    public static MultilingualE5SmallInternalServiceSettings.Builder fromPersistentMap(Map<String, Object> map) {
         ValidationException validationException = new ValidationException();
+        var requestFields = extractRequestFields(map, validationException);
+
+        Integer parsedDimensions = ServiceUtils.removeAsType(map, ServiceFields.DIMENSIONS, Integer.class);
+        if (parsedDimensions == null) {
```
Comment (jonathan-buttner) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/elasticsearch/MultilingualE5SmallInternalServiceSettings.java:81]: When the inference endpoint is created we will not allow the user to specify the dimensions. This code is for parsing persisted configuration.

**Comment by jonathan-buttner** [comment]
@elasticmachine merge upstream

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/elasticsearch/MultilingualE5SmallInternalServiceSettings.java (hunk 6 above)
Comment (davidkyle) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/elasticsearch/MultilingualE5SmallInternalServiceSettings.java:62]: Is there any point serialising or persisting these fields when they are immutable. dimensions cannot change and cosine similarity should be used for this model

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/elasticsearch/MultilingualE5SmallInternalServiceSettings.java [hunk 8]
```diff
@@ -61,25 +126,51 @@ public static MultilingualE5SmallInternalServiceSettings.Builder fromMap(Map<Str
             }
         }
 
+        SimilarityMeasure similarity = extractSimilarity(map, ModelConfigurations.SERVICE_SETTINGS, validationException);
```
Comment (davidkyle) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/elasticsearch/MultilingualE5SmallInternalServiceSettings.java:129]: Similarity should be cosine, if we allow the user to specify it (and align with the other model configs) then there should be a check that `similarity == cosine`

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/elasticsearch/MultilingualE5SmallInternalServiceSettings.java (hunk 6 above)
Comment (jonathan-buttner) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/elasticsearch/MultilingualE5SmallInternalServiceSettings.java:62]: Yeah probably not, I'll remove this and just force them to be `384` and `cosine`.

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/elasticsearch/MultilingualE5SmallInternalServiceSettings.java (hunk 8 above)
Comment (jonathan-buttner) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/elasticsearch/MultilingualE5SmallInternalServiceSettings.java]: Yeah good point, we'll just hardcode it to `cosine`.

**Review by davidkyle — APPROVED** [review]
LGTM

**Pull request merged**